### PR TITLE
support to distribute a folder alongwith its subfiles/subfolders

### LINKED ISSUE #394: [xparl] support distribution of all the files in the folder
XPARL doesn't support distributing a folder now. It would be more convenient if we support this feature.

### PR DESCRIPTION
**Pull request by YuechengLiu**
Empty subfolders are also distributed.

### CODE REVIEW

File: parl/remote/client.py [hunk 1]
```diff
@@ -104,15 +108,20 @@ def read_local_files(self, distributed_files=[]):
...
-                if not os.path.isdir(pathname):
+                if os.path.isdir(pathname):
+                    for root, dirs, files in os.walk(pathname):
+                        # distribute files in the folder
+                        if files:
+                            for sub_file in files:
+                                parsed_distributed_files.add(
+                                    os.path.join(root, sub_file))
+                        # dirtribute empty subfolders
+                        elif not dirs:
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/client.py]: not dirs -> len(dirs) == 0

Comment (YuechengLiu) [reply parl/remote/client.py]: modified

**Review by zenghsh3 — DISMISSED** [review]
LGTM

**Review by TomorrowIsAnOtherDay — DISMISSED** [review]

File: parl/remote/client.py [hunk 2]
```diff
@@ -113,14 +112,13 @@ def read_local_files(self, distributed_files=[]):
...
             for pathname in parsed_list:
                 if os.path.isdir(pathname):
-                    distribute_codefiles, distribute_otherfiles, distribute_emptysubfolders = get_subfiles_recursively(
+                    pythonfiles, otherfiles, emptysubfolders = get_subfiles_recursively(
                         pathname)
-                    otherfiles = otherfiles.union(distribute_codefiles,
-                                                  distribute_otherfiles)
-                    emptysubfolders = emptysubfolders.union(
-                        distribute_emptysubfolders)
+                    user_files += pythonfiles
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/client.py]: user_files.extend(pythonfiles)

File: parl/remote/client.py [hunk 3]
```diff
@@ -130,17 +128,17 @@ def read_local_files(self, distributed_files=[]):
             sep = os.sep
             if sep in main_file:
                 main_folder = sep.join(main_file.split(sep)[:-1])
-        codefiles = codefiles.union(
-            filter(lambda x: x.endswith('.py'), os.listdir(main_folder)))
+        code_files = filter(lambda x: x.endswith('.py'),
+                            os.listdir(main_folder))
 
-        for file_name in codefiles:
+        for file_name in set(code_files):
```
Comment (TomorrowIsAnOtherDay) [inline parl/remote/client.py]: It is unnecessary to use the set here.

File: parl/remote/client.py (hunk 2 above)
Comment (YuechengLiu) [reply parl/remote/client.py]: modified

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
LGTM.

**Pull request merged**